### ISSUE TITLE
Add telemetry probe for dex load time

### ISSUE THREAD

**Issue by ecsmyth** [mozilla-mobile/fenix#8803]
_edit:_ this issue was repurposed to only instrument dex load time instead of startup time but the description for startup probes remain below.

### Why/User Benefit/User Problem
Startup time is an oft cited reason for friction and churn. We need to understand how long it takes Fenix to startup for our users so that we can prioritize work to address regressions and outliers that result in decreased engagement or churn. Local and CI-based testing provides an incomplete picture of how Fenix performs at startup.

### Impact
Improved understanding of how Fenix performs on startup for our users.

### Acceptance Criteria (how do I know when I’m done?)
We should make a best-effort to _incrementally_ instrument as much of startup as possible - until we hit diminishing returns - because it's impossible to do it all. Unfortunately, it will be difficult to define what this looks like in advance so take this as a guiding principle. Msg mcomella if this does not make sense.

Some incremental steps we should try to implement for GA:
* Add a metric that tracks the start time of GV
* Add a metric that tracks the duration of `FenixApplication.onCreate` (this may or may not be helpful but we want to add it to see if it is in practice)
* Add a metric that tracks `FenixApplication.onCreate` until first frame drawn
* Distinguish between first run and other cold starts (don't let this impact performance!)
* (is this reasonable?) Distinguish between which page was shown, i.e. which startup path was used
* Time from first frame drawn to reportFullyDrawn? Essentially, capture loading top sites & open tabs because this is dependent on content quantity

Some incremental steps we probably can't implement by GA but we should verify:
* Add a probe that tracks `FenixApplication.onCreate` until visual completeness
* Add warm/hot start times, distinguish from cold start
* Distinguish between cold starts after upgrading and other cold starts

Here is ecsmyth's ideal startup time probe that inspired these incremental steps:
> Glean telemetry probe landed in Fenix release that measures time, as closely as is practical, from when user initiates startup until the app is started and the initial screen is visually complete. The probe should differentiate between hot, cold, and warm startup scenarios and account for differences in the first page rendered (e.g., how desktop measures first paint and first paint of about:home)

**Comment by mcomella** [comment]
Waiting on ecsmyth to determine if time for `GeckoRuntime.init` (i.e. the time Gecko starts on the main thread before continuing work on a background thread) is a valuable probe.

**Comment by mcomella** [comment]
> Waiting on ecsmyth to determine if time for `GeckoRuntime.init` (i.e. the time Gecko starts on the main thread before continuing work on a background thread) is a valuable probe.

I opened a new PR to add a probe I care about so I don't need this questionable one.

**Comment by mcomella** [comment]
I added the "dex launch time" probe: the time between process start and `Application.<init>`. I'll leave this issue open to remember to investigate the results and potentially file a follow-up for a more involved analysis.

**Comment by mcomella** [comment]
I'm going to close as fixed and create a new issue for the analysis.

**Comment by mcomella** [comment]
Filed https://github.com/mozilla-mobile/fenix/issues/10161 for the analysis.
